iNat Import

JoeCohen opened the pull request:
PR aims to implement the minimum acceptable functionality for NEMF for importing iNat observations
- Allows an MO user to import a list of that User's iNat observation(s) to MO via the iNatImport form.
- Implements the [Basic Inat Import milestone](https://github.com/MushroomObserver/mushroom-observer/milestone/3)

### Note
- It uses an inefficient way of giving users feedback about jobs. I didn't know how to do the js-Turbo 🪄 voodoo. I think that's enough for NEMF because I removed the UI for importing all a user's observations.
- It doesn't rate limit. I think that's OK for NEMF for the the same reason.

### Questions for reviewers
- Is more functionality needed?
- Appearance
- General organization
- Better names for classes, methods, variables, ex: ImportedInatObs
- Any ideas about shortening/dividing long classes/methods
- Any ideas about fixing ABC violations
- Any other criticism is welcome.

### Manual Test Protocol
- Create a sacrificial iNat observation, making sure it has lat/lng. I suggest also selecting Obscured Geoprivacy.
- In MO: Make sure SolidQueue is running, Create Observation, Import from iNat http://localhost:3000/observations/inat_imports/new
- Fill in the fields, using the iNat observation # of the sacrificial iNat obs, and Submit
Result: redirects to Observation Index (to be change to Inat Import Job Tracker)
- Wait 10 seconds, then refresh screen
Expected result: Index includes a new Observation
- Goto the new Observation. and review it. The Location should make sense. If the iNat obs had Obscured Geoprivacy, then MO coordinates should be hidden from the public.
- Goto/refresh the sacrificial iNat Obs
Expected result: 
Description includes imported to MO and date of the import; 
Observation Fields include Mushroom Observer URL linking to the newly created **Local** observation. 
(In the production environment, it links to the live db.)
- Delete the sacrificial iNat Obs

nimmolo commented:
Did you try adding the key to Rails credentials yet? It's very easy and the steps for decoding aand re-encoding the credentials are documented online.

JoeCohen commented:
Thanks for the follow up.
No. it looks easy. But for my case, I need something
additional/diferent which I was able to do without using credentials.
I might refactor later.

On Fri, Jun 28, 2024 at 7:35 PM andrew nimmo ***@***.***>
wrote:

> Did you try adding the key to Rails credentials yet? It's very easy and
> the steps for decoding aand re-encoding the credentials are documented
> online.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/2150#issuecomment-2197863316>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFCWERA3VNYSUQKB54TZJYMNVAVCNFSM6AAAAABH6TYQSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOJXHA3DGMZRGY>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

JoeCohen committed:
Delete redundant, indirect InatObs# obs private method

JoeCohen committed:
Revise controller workflow to reduce iNat API requests

- Gets multiple obss per request. For each request, gets pages; for each page gets obss
- Gets commoa separated list of obss
- Updates search obss mocks and InatObsTest accordingly

JoeCohen committed:
Fix test_create_inat_import_authorized

JoeCohen committed:
Update controller stubs

- Fixes controller test failuers

JoeCohen committed:
Conform InatPhotoTest to search mocks

- Fixes error in prior commit

JoeCohen committed:
Controller test importing multiple observations

JoeCohen committed:
Remove restrictions which prevent multiple imports

- Removes restriction in controller
- Removes now-unneeded tests, translation string

JoeCohen committed:
Add iNat UserName field to import form

JoeCohen committed:
Shorten test method names

JoeCohen committed:
Require iNat username in import form

JoeCohen committed:
Add inat_username column to InatImports

JoeCohen committed:
Fix typo in form

- Includes test which exposed the typo

JoeCohen committed:
Save inat_username before redirect to iNat

JoeCohen committed:
`post` (not `put`) to`create`

JoeCohen committed:
Prove that create saves InatImport.inat_username

JoeCohen committed:
Delete dead code

- Deletes obsolete `import_one_observation` method that made 1 iNatAPI request per observation

JoeCohen committed:
Add `user_login` param to iNat API request

- Prevents MO user from importing other people's iNat obss
(Currently doesn't completely prevent it; there's a workaround. But it can be prevented after authentication is functional.)

JoeCohen committed:
Fix line length violation

JoeCohen committed:
Fix bug where search has total_results: 0

JoeCohen committed:
Begin adding import all user's obs

- Add checkbox to form
- Add appropriate translation strings
- Require user to (a) check import all, and/or (b) provide a list of obss

JoeCohen committed:
Fix test_create_no_consent

- Assert specific flash text (vs just any old warning)
- Includes username in params
- Removes unnecessary code. (If there's no consent, it never calls the iNat API.)
- Trvial refacor of `create`

JoeCohen committed:
Add import_all column to InatImports

JoeCohen committed:
Mock result of iNat search for all a user's obss

JoeCohen committed:
change test method name

JoeCohen committed:
Test import all a user's obss

- Adds test
- Dynamically tweaks mock result to avoid (1) stubbing multiple iNat API requests, (2) stubbing photo imports
- Limits iNat API observation searches to Fungi and Protozo

NOTE: That last limit breaks other tests.

JoeCohen committed:
Improve local variable name for clarity

- mock_search => mock_search_result. It's the result of a search, not the search itself.

JoeCohen committed:
Refactor stub_inat_api_response for clarity

JoeCohen committed:
Add todo to Inat.rb

JoeCohen committed:
Add Protoza to stubbed iNat API requests

- constantizes `iconic_taxa`

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Define inat_obs#identifications

JoeCohen committed:
Add InatTaxon class

- Maps iNat :taxon to MO Name

JoeCohen committed:
Add Proposed Names to imported obss

JoeCohen committed:
Add comment explaining test expectation

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Trivial refactor of name_already_proposed?

JoeCohen committed:
Correct MO User for Naming

- Either the current user or the inat_manger

JoeCohen committed:
Improve add_inat_summmary_data

- extracts method
- uses `HEREDOC` for readability
- fixes `InatImportsController#add_inat_summmary_data` ABC size, but trades that for violation in extracted method

JoeCohen committed:
More detailed test assertions

JoeCohen committed:
Rewrite assetions for DRYness and clarity

JoeCohen committed:
Add Votes for Names corresponding to iNat identifications

JoeCohen committed:
Insure that Obs.name has Naming & positive vote

JoeCohen committed:
Test Namings where consensus has iNat Identification

JoeCohen committed:
Combine 2 test methods

- Reduces db hits

JoeCohen committed:
Add iNat Complexes

- Create an MO Naming if the iNat obs has an Identification of a Complex with an MO equivalent.
- Make the MO consensus a complex if the iNat Community ID is a Complex with an MO equivalent.

JoeCohen committed:
Extract method

- fixes ABC offenses

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' of https://github.com/MushroomObserver/mushroom-observer into jdc-1955-inat-import

JoeCohen committed:
Fix typo

JoeCohen committed:
Add InatObs provisional name methods

- `inat_provisional_name` gets a provisional species name observation field
- `provisional_name` maps it to an MO text_name

JoeCohen committed:
Fix InatObs#sequences bugs

- Stops this method from zapping `inat_obs_fields`
- Refines condition for text :datatype sequences
Text must start with 10 ACGT bases

JoeCohen committed:
Draft controller test for provisional names

JoeCohen committed:
remove stray breakpoint

JoeCohen committed:
Import prov name that exists in MO

JoeCohen committed:
`create` rather than `new` in test

JoeCohen committed:
Simple small refactor of iNatImportscontroller

- extracts `add_naming_with_vote` method
- `create` instead of `new` `save` combination
- `update` instead of `=` `save` combination
- Improves various ABC sizes

JoeCohen committed:
Clarify some method names

- Makes names more precise
- Adds clarifying comments

JoeCohen committed:
Update comment outlining import work flow

JoeCohen committed:
Add Name if iNat obs has a nom prov that's not in MO

- Adds test
- Uses MO API to add Name

JoeCohen committed:
Update Caveat/Details in import form

- Explains use of iNat Provisional Name

JoeCohen committed:
Add mock observation to test infrageneric names

JoeCohen committed:
Fix mapping of infrageneric names

- Maps iNat's rank+epithet to a full MO name.
E.g. "Section Distantes" to "Morchella sect. Distantes"

JoeCohen committed:
Fix infraspecific names

- Inserts rank between species and final epithet
- Privatizes methods used `by InatObs#name_id`
- Improves method name

JoeCohen committed:
reorganize InatObs private methods

nimmolo committed:
Suggestions for JWT

nimmolo committed:
Comments updated

nimmolo committed:
Rename token variables

nimmolo committed:
Fix RestClient.get, add URI.encode_www_form(query_args)

nimmolo committed:
Make @inat_import an ivar

JoeCohen committed:
Fix typo in iNat observation search param key

JoeCohen committed:
Fix stubbing request for access token

- Move token to response body field `access_token`
- Improve method name by specifying token type (access_token, not jwt).

JoeCohen committed:
Fix remaining errors in this branch

- Update stub of iNat API request
(conforms param order and format to those in this branch)
- Stub jwt request

JoeCohen committed:
Fix assertion failure message

JoeCohen committed:
change order of stubs

Stab in the dark to see if CI does any better.

JoeCohen committed:
Remove User-Agent param from stub_inat_api_request

Thought I removed it earlier.
I think this will fix the CI failure

JoeCohen committed:
Merge pull request #2307 from MushroomObserver/nimmo-jwt-suggestions

Suggestions for JWT

JoeCohen committed:
Minor clarification to InatImportController

Reminds me how this works
- Add some comments
- Better method name

JoeCohen committed:
inat import distantes

JoeCohen committed:
InatObsTest of obscured observation

Proves:
- iNat `:private_location` is imported
- gps data is hidden
- Location is blurred by iNat `public_position_accuracy`

JoeCohen committed:
Always get JWT after getting access token

Hopefully fixes push CI failure (intermittent?)
https://github.com/MushroomObserver/mushroom-observer/actions/runs/10433557690/job/28898401164?pr=2150
How did this ever pass? Guess: stubs persist among test methods.
One method's stub of jwt was used by another test.

JoeCohen committed:
Increase InatController coverage

Only 1 line uncovered. I will fix that when I refactor.

JoeCohen committed:
Refactor InatImportsControllerTest

- Refactors interactions with iNat
- Add method simulating all iNat interactions
- Refactors simulating iNat accreditations
  - Get authorization and tokens with 1 call
  - Do the double exchange with a 1 call
- Reorganizes helper methods

JoeCohen committed:
Delete stray commented-out code

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' of https://github.com/MushroomObserver/mushroom-observer into jdc-1955-inat-import

JoeCohen committed:
Small refactor of InatImportsController

- Better method and variable names
- Add comments re iNat documentation
- Misc. other changes

JoeCohen committed:
Improve outline of iNat import workflow

JoeCohen committed:
Fix blurred bounding box for obscured observations

It was about 2x too tall & wide.

JoeCohen committed:
Map iNat taxon to MO Name if many Names match

- Improves the mapping logic.
Previously it picked Fungi.
It's better to pick a Name if one is available.

JoeCohen committed:
generate job

JoeCohen committed:
DRY InatControllerTest constants

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' into jdc-inat-import-job

JoeCohen committed:
Begin draft of InatImportJob (still red)

- Change one controller test method to assert that job is enqueued
- Move the bulk of `authenticate` action to a job
- Move controller constants to job file

JoeCohen committed:
More iNatImportJob progress (still Errors)

- Fix some constants
- restore ivar in InatImportJob
- delete some commented out code
- Fix various references to user.

JoeCohen committed:
WIP: 4fcbe070f More iNatImportJob progress (still Errors)

JoeCohen committed:
Better name for redirect uri

`authenticate` => `authorization_response`
I've alwasy been uncomforatable with "authenticate".
- The action does either more than authentication (imports the Inat obss) or
less than authentication (if the user did not authorize MO to get confidential data)

JoeCohen committed:
Fix test calls to authorization_response action

JoeCohen committed:
Merge branch 'jdc-inat-import-job' of https://github.com/MushroomObserver/mushroom-observer into jdc-inat-import-job

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' into jdc-inat-import-job

(Changes REDIRECT_URI)

JoeCohen committed:
Reduce parameters of InatImportJob#perfom

- applies Introduce Parameter Object refactor

JoeCohen committed:
Do all authentication in background

- Moves remaining Controller authentication steps to ActiveJob
- Makes some method names more expressive
- Updates flow outline

JoeCohen committed:
Fix constant definition and reference thereto

JoeCohen committed:
Restore performing ActiveJobs in background

JoeCohen committed:
Mark controller tests to be moved to job test

- Adds TODO's, skips tests

JoeCohen committed:
skip another controller test

JoeCohen committed:
update documentation re mock search examples

JoeCohen committed:
Job test of simple observation

JoeCohen committed:
Delete duplicate method

JoeCohen committed:
Delete include of Rails url_helpers

- Hopefully fixes CI failure
See https://github.com/MushroomObserver/mushroom-observer/actions/runs/10497012210/job/29079910786
```ruby
ERROR InatImportJobTest#test_index_url (120.23s)
Minitest::UnexpectedError:         ArgumentError: Missing host to link to! Please provide the :host parameter, set default_url_options[:host], or set :only_path to true
            vendor/bundle/ruby/3.3.0/gems/actionpack-7.1.3.4/lib/action_dispatch/http/url.rb:64:in `full_url_for'
            vendor/bundle/ruby/3.3.0/gems/actionpack-7.1.3.4/lib/action_dispatch/http/url.rb:54:in `url_for'
            vendor/bundle/ruby/3.3.0/gems/actionpack-7.1.3.4/lib/action_dispatch/routing/route_set.rb:334:in `block in <class:RouteSet>'
            vendor/bundle/ruby/3.3.0/gems/actionpack-7.1.3.4/lib/action_dispatch/routing/route_set.rb:212:in `call'
            vendor/bundle/ruby/3.3.0/gems/actionpack-7.1.3.4/lib/action_dispatch/routing/route_set.rb:327:in `block in define_url_helper'
```

JoeCohen committed:
JobTest importing obs with 1 image

- Successfully stubs calls to `InatPhotoImporter`
  -  creates a `mock_photo_importer`mock object
  - mocks the :new method of that object
  - mocks the :api method of that object
  - stubs `InatPhotoImporter.new` to return that object

JoeCohen committed:
Refactor in light of new job test

- tweak some comments
- delete counterpart, skipped controller test

JoeCohen committed:
Move another test from ControllerTest to JobTest

JoeCohen committed:
Refactor pre-Job call part of JobTest methods

- Extract methods
- Eliminate unnecessary parameters
- Remove unused code

JoeCohen committed:
DRY assertions parts of JobTest methods

- Extracts standard assertions

JoeCohen committed:
Move rest of skipped controller tests to JobTest

- Move skipped controller methods (leaving them skipped)
- Delete unused controller test utility methods

JoeCohen committed:
Test import of sequenced observation

- Improves stub of photo importer to not fail if obs has multiple photos.

JoeCohen committed:
Test import of iNat obss with provisional names

JoeCohen committed:
JobTest of importing plant

JoeCohen committed:
JobTest when there are no matching iNat ids

JoeCohen committed:
JobTest importing multiple listed obss

JoeCohen committed:
Move test_import_all to JobTest

JoeCohen committed:
refactor add_inat_images to fix ABC offense

JoeCohen committed:
refactor import_one_result to fix ABC offense

JoeCohen committed:
Change vote for consensus to the max allowed

(per discussion with @mo-nathan)

JoeCohen committed:
Change vote for non-consensus to the max allowed

JoeCohen committed:
100% cover for InatImportsController

JoeCohen committed:
RuboCop InatImportJobTest

JoeCohen committed:
Fix Lint/UselessAssignment

JoeCohen committed:
Improve method name

JoeCohen committed:
Update Readme with some usernames

JoeCohen committed:
Clarify comment

JoeCohen committed:
Fix misspelt taxon

- Updates file name, readme, test that uses file

JoeCohen committed:
Fix bug re importing obss with nom. prov.

- Adds test to prevent reversion
- Fixes bug, changing order of method defs for easier reading
- Augments all JobTests to prove that obs was added

JoeCohen committed:
Test importing obs with infra-specific name

JoeCohen committed:
Delete stray breakpoint

JoeCohen committed:
Fix importing Complexes

- Add InatObs test, fix InatObs mapping of Complex to MO Name
- Add Job Test

JoeCohen committed:
Merge pull request #2329 from MushroomObserver/jdc-inat-import-job

Background iNat imports

JoeCohen committed:
Tweak mapping of iNat "Complex"

JoeCohen committed:
Document issues with iNat projects

JoeCohen committed:
Better name for method

JoeCohen committed:
JobTest exposing 0% Community Vote bug

- See #2336

JoeCohen committed:
Update VoteCache by recalculating consensus

- Fixes #2336, but introduces a new bug:
The consensus is now wrong.
That could be fixed by increasing the precision of Namings.created_at

JoeCohen committed:
Fix bug in test

- Exposes new bug: consensus intermittently incorrect

JoeCohen committed:
Fix bug exposed by prior commit

- The Name which I last explicitly `Create` is not necessarily `Name.last`.
E.g., MO implicitly creates a genus if I explictly `Create` a species (or group) whose genus is not yet in MO.

JoeCohen committed:
Add flash that import has started

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Strip inat_username that was input by user

JoeCohen committed:
Clearer Observation Detail line for iNat imports

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Remove form option to import all observations

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
migrate and update schema

JoeCohen committed:
Remove obsolete, now-incorrect assert

JoeCohen committed:
Delete Gemfile.lock and run bundle install

Hoping this fixes [CI failure](https://github.com/MushroomObserver/mushroom-observer/actions/runs/10752167118/job/29820853988)
(Perhaps Gemfile.lock was "corrupted". Cf. https://github.com/rubygems/rubygems/issues/6124.)

nimmolo committed:
Inflection `INat`, subclass POROs

nimmolo committed:
Revert inflection

JoeCohen committed:
Restore `Import from iNat` tab to Create Observation page

- Adds assertion to prevent reversion
- Adds new-style tab
- Updates "translation" strings and names

JoeCohen committed:
Add change carelessly omitted from prior commit

JoeCohen committed:
Fix test assertion

- Narrows the assertion so that it passes only on the right page.

JoeCohen committed:
Update flash error when list of iNat id's is missing

JoeCohen committed:
Prevent import of previously imported iNat obs

JoeCohen committed:
Warn of import of iNat obs mirrored from MO

JoeCohen committed:
Try to fix markdownlint Codeclimate issues in README

JoeCohen committed:
Try to turn off markdownlint linelength rule for tables

- Applies Copilot solution:
>To disable the line length rule for tables in markdownlint when using Codeclimate, you can configure the .markdownlint.json file. Here’s how you can do it:
>
> Create or update the .markdownlint.json file in your project’s root directory.
Add the following configuration to disable the line length rule (MD013) specifically for tables:
```json
{
  "MD013": {
    "tables": false
  }
}
```

JoeCohen committed:
Fix one (false) TODO violation

JoeCohen committed:
Second try at disabling markdownint linelength for tables

- Removes ineffective .markdownlint.json file
- Try to do it via .codeclimate.yml

JoeCohen committed:
Undo failed disable markdownlint rule via .codeclimate.yml

JoeCohen committed:
Rubocopping

JoeCohen committed:
Replace code TODO's from code with GitHub Issues

- Removes `TODO`s from the code base, replacing them with Issues in the GitHub [iNat Import Project](https://github.com/orgs/MushroomObserver/projects/6)

JoeCohen committed:
Change more TODO's FIXME's to GitHub issues

JoeCohen committed:
Cross-reference GitHub Issues in NOTE's

JoeCohen committed:
Add inat_obs#source

- Define `source` in `InatObs` rather than job
- Fixes Codeclimate Fixme violation

JoeCohen committed:
Delete FIXME which was fixed in prior commit

JoeCohen committed:
Delete redundant TODO

JoeCohen committed:
Fix line length violation

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Delete line accidentally added to schema

JoeCohen committed:
Update schema after migration

- Includes both main and local changes

JoeCohen committed:
Complete the iNat `Mushroom Observer URL` observation field

- Adds and completes iNat's `Mushroom Observer URL` observation field when importing an observation

JoeCohen committed:
Stub adding observation field to inat obaservation

- Fixes test Errors of prior commit

JoeCohen committed:
Update iNat obs description when importing to MO

- Adds "Imported to Mushroom Observer yyyy-mm-dd" to the description

JoeCohen committed:
Fix prior commit test Failures

- fixes `inat_description`
- fixes stub of updating description

JoeCohen committed:
Improve some labels and titles

JoeCohen committed:
Make controller (view) tests less brittle

- Fixes fialures caused by prior commit.

JoeCohen committed:
List un-imported fields in Snapshot

JoeCohen committed:
Multi-environment credentials for iNat

- Facilitates testing

JoeCohen committed:
Undo prior commit

JoeCohen committed:
Fix typo in credential

JoeCohen committed:
Update schema from current live db snapshot

- Downloaded most recent live db snapshot, migrated, commit changes

JoeCohen committed:
Add InatImporJobTrackers

- Generate controllers and model
- Run migration
- Add InatImportJobTracker#show route

JoeCohen committed:
Delete skipped test of importing projects

Importing projects is low priority and won't be tackled now.
A test should be added when/if implementing #2245.

JoeCohen committed:
Rename column

JoeCohen committed:
Rename column in fixtures

JoeCohen committed:
Send user to job tracker if importing iNat obs

- Goes to tracker instead of Observation index

JoeCohen committed:
Define InatImport enum with positional arguments

- Fixes deprecation warning

JoeCohen committed:
Improve fixture

JoeCohen committed:
Update InatImport associations

JoeCohen committed:
Add class comment

JoeCohen committed:
Reverse changes to associations

- Fixes this ERROR
```ruby
ERROR Observations::InatImportsControllerTest#test_create_authorization_request (0.54s)
Minitest::UnexpectedError:         ArgumentError: You tried to define an enum named "state" on the model "InatImport", but this will generate a instance method "Unstarted?", which is already defined by another enum.
            app/models/inat_import.rb:15:in `<class:InatImport>'
            app/models/inat_import.rb:14:in `<top (required)>'
```

Maybe this was cuasesd by naming the InatImportJobTrackers column `inat_import` instead of `inat_import_id`.
I don't have time to think about it now.

JoeCohen committed:
Skeleton tests of InatImportJobTracker

JoeCohen committed:
Fix prior commit failures

JoeCohen committed:
Add some more stuff on tracker page

JoeCohen committed:
id's for tracker show page elements

So that they can eventually be updated by js

JoeCohen committed:
Add InatImport attributes for non-js tracking

Adds two attributes each to be:
- updated by the job
- read by the job tracker

JoeCohen committed:
Add changes accidentally omitted from prior commit

JoeCohen committed:
Improve msg left in description of imported iNat obss

JoeCohen committed:
Update tracker#show with # of importable obss

JoeCohen committed:
Increment count of imported obss

JoeCohen committed:
Your Observations & Update Resuilts tracker#show buttons

JoeCohen committed:
Restore enqueing of InatImportJobs

Oops

JoeCohen committed:
Handle error thrown by call to iNat API

- Rescue error
- Return when page of results is just an error.
- Allows job to proceed to Done
- Fixes #2395

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

- Resolves latest set of conflicts.

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' into nimmo-inat-inflection

- Resolves conflicts

JoeCohen committed:
Update subclasses with more recent changes from old class

JoeCohen committed:
Merge pull request #2372 from MushroomObserver/nimmo-inat-inflection

`Inat` subclass POROs

JoeCohen committed:
Delete now-redundant flash success message

JoeCohen committed:
Delete redundant, useless footer

JoeCohen committed:
Fix typo in comment

JoeCohen committed:
improve method name

JoeCohen committed:
Make inat_obs an ivar in InatImportJob

- Tired of passing it as a method arg.

JoeCohen committed:
"none" instead of blank for Obs Fields if there are none

See #2387

JoeCohen committed:
Finish Sequence & Observation Fields snapshot cleanup

- Translate "none"
- Delete explicit Sequences heading in Snapshot. It's redundant because they are all iNat Observation Fields

JoeCohen committed:
Reformat line causing "undefined tag reference" Failure

- Magically cures mysterious failure

JoeCohen committed:
Remove stray close brace at end of Snapshot Comment

mo-nathan committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Fix disappearing iNat photo bug

JoeCohen committed:
Merge branch 'jdc-1955-inat-import' of https://github.com/MushroomObserver/mushroom-observer into jdc-1955-inat-import

JoeCohen committed:
Conform stub to the fix of the prior commit

(Prior commit changed `:put` to :`patch`)

mo-nathan commented:
It's working well.  It doesn't seem like it should need me to type my iNat username every time.  Seems like this could be saved in the MO database after the first authentication cycle.

mo-nathan commented:
Since the Turbo/JS magic isn't updating the page, I think the page should say something about manually refreshing the page.  Maybe after the "Status" if the status isn't "Done", it could say "Refresh to monitor the status" or something along those lines.

JoeCohen commented:
How about instead changing the name of the button?
> Since the Turbo/JS magic isn't updating the page, I think the page should say something about manually refreshing the page. Maybe after the "Status" if the status isn't "Done", it could say "Refresh to monitor the status" or something along those lines.

mo-nathan commented on app/jobs/inat_import_job.rb:
```diff
@@ -0,0 +1,408 @@
+    increment_imported_count
+  end
+
+  def create_observation
```
For this ABC metric violation, I consider implementing something like `create_hash` on the the `Inat::Obs` that returns a hash which you could then add in the other stuff with a merge and pass the resulting to `create` rather than the big long list of key value pairs.

JoeCohen committed:
Display iNat API observation search errors in tracker#show

- Adds InatImport `response_errors` column
- Writes that column in the job if the iNat API retuns a bad response
- Reads the column in the job tracker show page

JoeCohen committed:
Fix id of span which contains text of errors

mo-nathan commented on app/jobs/inat_import_job.rb:
```diff
@@ -0,0 +1,408 @@
+    Comment.create(params)
+  end
+
+  def import_snapshot
```
For this ABC metric violationm I'd move most of this into the `InatObs` object as a `comment` method or some such.  That would probably go a long way towards addressing this since you'd no longer be doing all those dereferences.

mo-nathan commented on app/controllers/observations/inat_imports_controller.rb:
```diff
@@ -0,0 +1,190 @@
+
+    def new; end
+
+    def create
```
For this ABC metrics violation, I would start with moving all the error checking into it's own method so it could be a single lline in this method.

mo-nathan commented on app/classes/inat/obs.rb:
```diff
@@ -0,0 +1,463 @@
+
+    ########## iNat attributes
+
+    def inat_description
```
I'd start by loking for a way of getting rid of this pattern,  This seems very repetitive.

JoeCohen committed:
Error handling & display for all iNat API calls

- Rescues the errors
- Displays them on the job tracker page

JoeCohen commented:
@mo-nathan: Thanks so much for looking at this and for all the suggestions.
I'll work on them tomorrow (after dealing with the new conflicts).

JoeCohen committed:
Error handling for remaining external calls in job

JoeCohen committed:
Revise job tracker show page

- Prominently display refresh button if job is incomplete
- Alway display Your Observations button if job is done (including when there are errors).

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen commented:
@mo-nathan Here's what I did to the tracker page. I don't have strong feelings about this, so am happy to make any other changes you'd like. Some options:
- Change the button name to  "Refresh to monitor the status", "Click to update status" or anything else.
- A flash  like "Import started. Click the Update button to update status"when the user (first) lands on the page.
- Additional text like what you proposed.

<img width="304" alt="Screenshot 2024-09-17 at 2 37 22 PM" src="https://github.com/user-attachments/assets/1ef19de3-72c9-445b-b7c5-999ecb8f72cb">

> Since the Turbo/JS magic isn't updating the page, I think the page should say something about manually refreshing the page. Maybe after the "Status" if the status isn't "Done", it could say "Refresh to monitor the status" or something along those lines.

JoeCohen committed:
Save inat_username draft

JoeCohen committed:
Fix updating the iNat observation `description`

- PUT instead of PATCH but also
- `ignore_photos: 1`
See https://www.inaturalist.org/pages/api+reference#put-observations-id
https://forum.inaturalist.org/t/api-modify-observation-description/55665

JoeCohen committed:
Tweak when MO inat_username is updated

- Update it if the job is successful enough to believe that inat_username likely belongs to the current MO user.

JoeCohen committed:
Prevent a user from using another's iNat username

- Prevents user from importing someone else's iNat observation
- Prevents user from "stealing" someone else's iNat username
- Prevents user from accidentally changing his own`User.inat_username`

JoeCohen committed:
Put Errors below Status on tracker sho page

JoeCohen committed:
Fix issues with checking for username match

- Fixes finding mismatch
- Displays plain-language message if a mismatch
- Prevent stub persistence in `inat_import_job_test`

JoeCohen committed:
Enable InatImportJobTest test_user_name_mismatch

- Fixes bugs in test and un-Skips it.
- Adds comment explaining setup and teardown

JoeCohen committed:
Remove disabling (and enabling) of Webmock

Seems unnecessary. Also hoping it fixes the stub-related CI failures.

JoeCohen commented:
Saving and auto-filling the iNat username is implemented as of Commiit #[276338c](https://github.com/MushroomObserver/mushroom-observer/pull/2150/commits/276338c2033d4e8a948cf795fe952de096f55060)

JoeCohen committed:
Merge branch 'main' into jdc-1955-inat-import

JoeCohen committed:
Migrate to fix missing column error

JoeCohen committed:
Remove User-Agent from stub

Hoping this time it fixes CI failures.
Also conforms stub url to other calls to iNat API

JoeCohen committed:
Fix url for /users/me stub

JoeCohen committed:
easy Rubocop fixes, update flow documentation

JoeCohen committed:
Move snapshot content from job to model

- Implements @mo-nathan's suggestion at https://github.com/MushroomObserver/mushroom-observer/pull/2150#discussion_r1762032235

JoeCohen committed:
Fix ABC offense

- Extracts method

JoeCohen committed:
Primitive rate limiting

- Adds 1 sec delay between 2 iNat API requests that are very close to each other in order to comply with iNat Policy.
https://www.inaturalist.org/pages/developers#terms-of-use

JoeCohen committed:
Turn off markdownlint line length limits for tables

- Tries to fix Codeclimate Line length offense

JoeCohen committed:
Fix `snapshot` ABC offense

- Replaces a string with many interpolated values with iterating over a hash.
